src/server/journal/serializer.cc

```diff
@@ -0,0 +1,204 @@
+    auto exp_val = (expr);                     \
+    if (!exp_val) {                            \
+      VLOG(1) << "Error while calling " #expr; \
+      return exp_val.error();                  \
+    }                                          \
+    dest = exp_val.value();                    \
+  } while (0)
+
+#define SET_OR_UNEXPECT(expr, dest)            \
+  {                                            \
+    auto exp_res = (expr);                     \
+    if (!exp_res)                              \
+      return make_unexpected(exp_res.error()); \
+    dest = std::move(exp_res.value());         \
+  }
+
+namespace dfly {
+
+JournalWriter::JournalWriter(io::Sink* sink, std::optional<DbIndex> dbid)
+    : sink_{sink}, cur_dbid_{dbid} {
+}
+
+error_code JournalWriter::WriteU8(uint8_t v) {
+  return sink_->Write(io::Bytes{&v, 1});
+}
+
+error_code JournalWriter::WriteU16(uint16_t v) {
+  uint8_t buf[2];
+  absl::big_endian::Store16(&buf, v);
+  return sink_->Write(io::Bytes{buf, sizeof(buf)});
+}
+
+error_code JournalWriter::WriteU64(uint64_t v) {
+  uint8_t buf[8];
+  absl::big_endian::Store64(&buf, v);
+  return sink_->Write(io::Bytes{buf, sizeof(buf)});
+}
+
+error_code JournalWriter::Write(std::string_view sv) {
+  RETURN_ON_ERR(WriteU64(sv.size()));
+  return sink_->Write(io::Buffer(sv));
+}
+
+error_code JournalWriter::Write(std::pair<std::string_view, ArgSlice> args) {
+  auto [cmd, tail_args] = args;
+
+  RETURN_ON_ERR(WriteU64(1 + tail_args.size()));
+  RETURN_ON_ERR(Write(cmd));
+  for (auto v : tail_args)
+    RETURN_ON_ERR(Write(v));
+
+  return std::error_code{};
+}
+error_code JournalWriter::Write(CmdArgList args) {
+  RETURN_ON_ERR(WriteU64(args.size()));
+  for (auto v : args)
+    RETURN_ON_ERR(Write(facade::ToSV(v)));
+
+  return std::error_code{};
+}
+
+error_code JournalWriter::Write(std::monostate) {
+  return std::error_code{};
+}
+
+error_code JournalWriter::Write(const journal::EntryNew& entry) {
+  if (entry.opcode != journal::Op::SELECT && (!cur_dbid_ || entry.dbid != *cur_dbid_)) {
+    RETURN_ON_ERR(Write(journal::EntryNew{journal::Op::SELECT, entry.dbid}));
+    cur_dbid_ = entry.dbid;
+  }
+
+  RETURN_ON_ERR(WriteU8(uint8_t(entry.opcode)));
+
+  switch (entry.opcode) {
+    case journal::Op::SELECT:
+      return WriteU16(entry.dbid);
+    case journal::Op::VAL:
+      RETURN_ON_ERR(WriteU64(entry.txid));
+      return std::visit([this](const auto& payload) { return Write(payload); }, entry.payload);
+    default:
+      break;
+  };
+  return std::error_code{};
+}
+
+JournalReader::JournalReader(io::Source* source, DbIndex dbid)
+    : source_{source}, buf_{}, dbid_{dbid} {
+}
+
+io::Result<uint8_t> JournalReader::ReadU8() {
+  uint8_t buf[1];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 1)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
```

Comment:
And sharing common error codes

However this would make this PR bigger than intended 🤔